validation.js

```diff
@@ -0,0 +1,281 @@
+  ) {
+    // prettier-ignore
+    return new Error('invalid message: string content must end with .box or .box2, on feed: ' + msgVal.author);
+  }
+  if (typeof content === 'object') {
+    if (!content.type || typeof content.type !== 'string') {
+      // prettier-ignore
+      return new Error('invalid message: content must have type, on feed: ' + msgVal.author);
+    }
+    if (content.type !== 'metafeed/index') {
+      // prettier-ignore
+      return new Error('invalid message: content type must be exactly "metafeed/index" on feed: ' + msgVal.author);
+    }
+    if (!Ref.isMsg(content.indexed)) {
+      // prettier-ignore
+      return new Error('invalid message: content must have indexed as a classic message ID, on feed: ' + msgVal.author);
+    }
+  }
+}
+
+function validateHmac(hmacKey) {
+  if (!hmacKey) return
+  if (typeof hmacKey !== 'string' && !Buffer.isBuffer(hmacKey)) {
+    return new Error('invalid hmac key: must be a string or buffer')
+  }
+  const bytes = Buffer.isBuffer(hmacKey)
+    ? hmacKey
+    : Buffer.from(hmacKey, 'base64')
+
+  if (typeof hmacKey === 'string' && bytes.toString('base64') !== hmacKey) {
+    return new Error('invalid hmac')
+  }
+
+  if (bytes.length !== 32) {
+    return new Error('invalid hmac, it should have 32 bytes')
+  }
+}
+
+function validatePayload(msgVal, payload) {
+  if (!payload) return
+  const expected = classic.getMsgId(payload)
+  const actual = msgVal.content.indexed
+  if (expected !== actual) {
+    return new Error('invalid message: payload msg key does not match')
+  }
+}
+
+function validateAsJSON(msgVal) {
+  const asJson = JSON.stringify(msgVal, null, 2)
+  if (asJson.length > 8192) {
+    // prettier-ignore
+    return new Error('invalid message: message is longer than 8192 latin1 codepoints');
+  }
+}
+
+function validateSync(nativeMsg, prevNativeMsg, hmacKey) {
+  let err
+  if ((err = validateShape(nativeMsg))) return err
+
+  const msgVal = Array.isArray(nativeMsg) ? nativeMsg[0] : nativeMsg
+  const payload = Array.isArray(nativeMsg) ? nativeMsg[1] : null
+
+  if ((err = validateHmac(hmacKey))) return err
+  if ((err = validateAuthor(msgVal))) return err
+  if ((err = validateHash(msgVal))) return err
+  if ((err = validateTimestamp(msgVal))) return err
+  if (prevNativeMsg) {
+    if ((err = validatePrevious(msgVal, prevNativeMsg))) return err
+    if ((err = validateSequence(msgVal, prevNativeMsg))) return err
+  } else {
+    if ((err = validateFirstPrevious(msgVal))) return err
+    if ((err = validateFirstSequence(msgVal))) return err
+  }
+  if ((err = validateOrder(msgVal))) return err
+  if ((err = validateContent(msgVal))) return err
+  if ((err = validatePayload(msgVal, payload))) return err
+  if ((err = validateAsJSON(msgVal))) return err
+  if ((err = validateSignature(msgVal, hmacKey))) return err
+}
+
+function validate(nativeMsg, prevNativeMsg, hmacKey, cb) {
+  let err
+  if ((err = validateSync(nativeMsg, prevNativeMsg, hmacKey))) {
```

Comment:
I wonder if we should at least validate that the payload message key is equal to indexed on index message?

Reply:
I think we do that! Check validatePayload

Reply:
Ah yes, thanks :) It was a bit hard to see from all the existing classic validation